tenders/hvt/hvt_cpu_x86_64.c

```diff
@@ -48,23 +48,47 @@ void hvt_x86_setup_pagetables(uint8_t *mem, size_t mem_size)
     uint64_t *pml4 = (uint64_t *)(mem + X86_PML4_BASE);
     uint64_t *pdpte = (uint64_t *)(mem + X86_PDPTE_BASE);
     uint64_t *pde = (uint64_t *)(mem + X86_PDE_BASE);
+    uint64_t *pte0 = (uint64_t *)(mem + X86_PTE0_BASE);
     uint64_t paddr;
 
     /*
      * For simplicity we currently use 2MB pages and only a single
      * PML4/PDPTE/PDE.  Sanity check that the guest size is a multiple of the
-     * page size and will fit in a single PDE (512 entries).
+     * page size and will fit in a single PDE (512 entries). Additionally,
+     * check that the guest size is at least 2MB.
      */
     assert((mem_size & (X86_GUEST_PAGE_SIZE - 1)) == 0);
     assert(mem_size <= (X86_GUEST_PAGE_SIZE * 512));
+    assert(mem_size >= X86_GUEST_PAGE_SIZE);
 
     memset(pml4, 0, X86_PML4_SIZE);
     memset(pdpte, 0, X86_PDPTE_SIZE);
     memset(pde, 0, X86_PDE_SIZE);
+    memset(pte0, 0, X86_PTE_SIZE);
 
     *pml4 = X86_PDPTE_BASE | (X86_PDPT_P | X86_PDPT_RW);
     *pdpte = X86_PDE_BASE | (X86_PDPT_P | X86_PDPT_RW);
-    for (paddr = 0; paddr < mem_size; paddr += X86_GUEST_PAGE_SIZE, pde++)
+
+    /*
+     * PDE[0] is special: use 4kB pages and PTE0 for the first 2MB of address
+     * space.
+     */
+    *pde = X86_PTE0_BASE | (X86_PDPT_P | X86_PDPT_RW);
+    pde++;
+    for (paddr = 0; paddr < X86_GUEST_PAGE_SIZE; paddr += 0x1000, pte0++)
+    {
+        /*
+         * Leave all pages below X86_PTE0_MAP_START not-present in the guest.
+         *
+         * This includes the zero page, boot GDT and the guest's page tables
+         * themselves.
+         */
+        if (paddr < X86_PTE0_MAP_START)
+            continue;
+        *pte0 = paddr | (X86_PDPT_P | X86_PDPT_RW);
```

Comment:
Good point. This will also enforce the documented API (https://github.com/Solo5/solo5/blob/master/include/solo5/solo5.h#L94) which explicitly says "The application MUST NOT make any further assumptions about memory layout".